python/gtsam/examples/Pose3ISAM2Example.py

```diff
@@ -0,0 +1,181 @@
+
+    axes.set_xlim3d(-30, 45)
+    axes.set_ylim3d(-30, 45)
+    axes.set_zlim3d(-30, 45)
+    plt.pause(1)
+
+    return marginals
+
+
+def createPoses():
+    """
+    Creates ground truth poses of the robot.
+    """
+    P0 = np.array([[1, 0, 0, 0],
+                   [0, 1, 0, 0],
+                   [0, 0, 1, 0],
+                   [0, 0, 0, 1]])
+    P1 = np.array([[0, -1, 0, 15],
+                   [1, 0, 0, 15],
+                   [0, 0, 1, 20],
+                   [0, 0, 0, 1]])
+    P2 = np.array([[cos(pi/4), 0, sin(pi/4), 30],
+                   [0, 1, 0, 30],
+                   [-sin(pi/4), 0, cos(pi/4), 30],
+                   [0, 0, 0, 1]])
+    P3 = np.array([[0, 1, 0, 30],
+                   [0, 0, -1, 0],
+                   [-1, 0, 0, -15],
+                   [0, 0, 0, 1]])
+    P4 = np.array([[-1, 0, 0, 0],
+                   [0, -1, 0, -10],
+                   [0, 0, 1, -10],
+                   [0, 0, 0, 1]])
+    P5 = P0[:]
+
+    return [Pose3(P0), Pose3(P1), Pose3(P2), Pose3(P3), Pose3(P4), Pose3(P5)]
+
+def Pose3_ISAM2_example():
+    """
+    Perform 3D SLAM given ground truth poses as well as simple
+    loop closure detection.
+    """
+    plt.ion()
+
+    def vector6(x, y, z, a, b, c):
+        """Create 6d double numpy array."""
+        return np.array([x, y, z, a, b, c], dtype=float)
+
+    # Although this example only uses linear measurements and Gaussian noise models, it is important
+    # to note that iSAM2 can be utilized to its full potential during nonlinear optimization. This example
+    # simply showcases how iSAM2 may be applied to a Pose2 SLAM problem.
+    PRIOR_NOISE = gtsam.noiseModel.Diagonal.Sigmas(vector6(0.1, 0.1, 0.1, 0.3, 0.3, 0.3))
+    ODOMETRY_NOISE = gtsam.noiseModel.Diagonal.Sigmas(vector6(0.1, 0.1, 0.1, 0.2, 0.2, 0.2))
+
+    # Create the ground truth poses of the robot trajectory.
+    poses = createPoses()
+
+    # iSAM2 parameters which can adjust the threshold necessary to force relinearization and how many
+    # update calls are required to perform the relinearization.
+    parameters = gtsam.ISAM2Params()
+    parameters.setRelinearizeThreshold(0.1)
+    parameters.setRelinearizeSkip(1)
+    isam = gtsam.ISAM2(parameters)
+
+    # Create a Nonlinear factor graph as well as the data structure to hold state estimates.
+    graph = gtsam.NonlinearFactorGraph()
+    initial_estimate = gtsam.Values()
+
+    # Add prior factor to the first pose with intentionally poor initial estimate
+    graph.push_back(gtsam.PriorFactorPose3(X(0), poses[0], PRIOR_NOISE))
+    initial_estimate.insert(X(0), poses[0].compose(gtsam.Pose3(
+        gtsam.Rot3.Rodrigues(-0.1, 0.2, 0.25), gtsam.Point3(0.05, -0.10, 0.20))))
+
+    def determine_loop_closure(odom, current_estimate, xyz_tol=0.6, rot_tol=0.3):
+        """
+        Simple brute force approach which iterates through previous states
+        and checks for loop closure.
+        Author: Jerred Chen
+        ### Parameters:
```

Comment:
very cool, thanks for this PR. Usually we follow the Google style for docstrings which states:
https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

```python
Args:
    odom: 1x6 vector representing noisy odometry transformation ...
    current_estimate: The current estimates computed by iSAM2.
    ...

Returns:
```

A python GTSAM example is here:
https://github.com/borglab/gtsam/blob/develop/python/gtsam/examples/TranslationAveragingExample.py#L99